src/psyclone/transformations.py

```diff
@@ -340,67 +379,124 @@ def name(self):
+            node1_fs_name = node1.field_space.orig_name
+            node2_fs_name = node2.field_space.orig_name
+            # 2.1) Check that both function spaces are valid
+            if not (node1_fs_name in VALID_FUNCTION_SPACE_NAMES and
+                    node2_fs_name in VALID_FUNCTION_SPACE_NAMES):
+                raise TransformationError(
+                    "Error in {0} transformation: One or both function "
+                    "spaces '{1}' and '{2}' have invalid names.".
+                    format(self.name, node1_fs_name, node2_fs_name))
+            # Check whether any of the spaces is ANY_SPACE. Loop fusion over
+            # ANY_SPACE is allowed only when the 'same_space' flag is set
+            node_on_any_space = node1_fs_name in VALID_ANY_SPACE_NAMES or \
+                node2_fs_name in VALID_ANY_SPACE_NAMES
+            # 2.2) If 'same_space' is true check for at least one ANY_SPACE
+            if same_space:
+                if not node_on_any_space:
                     raise TransformationError(
-                        "Error in DynamoLoopFuse transformation. "
-                        "Cannot fuse loops that are over different spaces: "
-                        "{0} {1}.{2}".format(node1.field_space.orig_name,
-                                             node2.field_space.orig_name,
-                                             info))
-            else:  # one or more of the function spaces is any_space
-                if not same_space:
+                        "Error in {0} transformation: The 'same_space' "
+                        "flag was set, but does not apply because "
+                        "neither field is on 'ANY_SPACE'.".format(self))
+            # 2.3) If 'same_space' is not True then make further checks
+            else:
+                # 2.3.1) Check whether one or more of the function spaces
+                # is ANY_SPACE without the `same_space` flag
+                if node_on_any_space:
                     raise TransformationError(
-                        "DynamoLoopFuseTrans. One or more of the iteration "
-                        "spaces is unknown ('any_space') so loop fusion might "
-                        "be invalid. If you know the spaces are the same then "
-                        "please set the 'same_space' optional argument to "
-                        "True.")
+                        "Error in {0} transformation: One or more of the "
+                        "iteration spaces is unknown ('ANY_SPACE') so loop "
+                        "fusion might be invalid. If you know the spaces "
+                        "are the same then please set the 'same_space' "
+                        "optional argument to 'True'.".format(self.name))
+                # 2.3.2) Check whether specific function spaces are the
+                # same. If they are not, the loop fusion is still possible
+                # but only when both function spaces are discontinuous
```

Comment:
I don't believe that discontinuous means we can fuse (or does it?)

Reply:
See the above comment about user-supplied kernels looping over cells (safe to fuse) and not dofs (not safe to fuse).